src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/meta/ConstantPool.java

```diff
@@ -165,6 +165,23 @@ interface BootstrapMethodInvocation {
         /**
          * Gets the static arguments with which the bootstrap method will be invoked.
          *
+         * An argument of type {@link PrimitiveConstant} represents a {@code CONSTANT_Dynamic_info}
```

Comment:
Can it not be an index to any valid LDC constant? This is implied by this code in your Graal PR:
```
if (constant instanceof PrimitiveConstant primitiveConstant) {
    int argCpi = primitiveConstant.asInt();
    Object argConstant = lookupConstant(argCpi, opcode == Opcodes.INVOKEDYNAMIC ? Opcodes.LDC : opcode, allowBootstrapMethodInvocation);
```

Reply:
Oh yes, sorry, I will fix this documentation.

Reply:
From the user point of view, a `PrimitiveConstant` can in fact only be a `ConstantDynamic`, because the other type of constant are automatically resolved in `CachedBSMArgs.get`. In the Graal PR, `lookupConstant` is a recursive call, allowing to call the associated bootstrap method.